File: src/main/java/com/example/hexagonal/application/command/CancelOrderHandler.java
package com.example.hexagonal.application.command;

import com.example.hexagonal.application.dto.CancelOrderOutput;
import com.example.hexagonal.application.port.input.CancelOrderUseCase;
import com.example.hexagonal.application.port.output.EventPublisher;
import com.example.hexagonal.application.port.output.OrderCommandRepository;
import com.example.hexagonal.domain.exception.OrderNotFoundException;

import java.util.ArrayList;

public class CancelOrderHandler implements CancelOrderUseCase {

    private final OrderCommandRepository orderRepository;
    private final EventPublisher eventPublisher;

    public CancelOrderHandler(OrderCommandRepository orderRepository,
                              EventPublisher eventPublisher) {
        this.orderRepository = orderRepository;
        this.eventPublisher = eventPublisher;
    }

    @Override
    public CancelOrderOutput execute(String orderId, String reason) {
        var order = orderRepository.findById(orderId)
                .orElseThrow(() -> new OrderNotFoundException(orderId));

        order.cancel(reason);

        orderRepository.save(order);
        eventPublisher.publishAll(new ArrayList<>(order.getDomainEvents()));
        order.clearEvents();

        return new CancelOrderOutput(
                order.getId(),
                order.getStatus().name(),
                reason);
    }
}

File: src/application/commands/CancelOrderHandler.ts
// ============================================================
// Command Handler - 取消訂單
// ============================================================

import {
  CancelOrderUseCase,
  CancelOrderOutput,
} from '../ports/input/CancelOrderUseCase';
import { OrderCommandRepository } from '../ports/output/OrderRepository';
import { EventPublisher } from '../ports/output/EventPublisher';
import { OrderNotFoundError } from '../../domain/errors/DomainError';

export class CancelOrderHandler implements CancelOrderUseCase {
  constructor(
    private readonly orderRepository: OrderCommandRepository,
    private readonly eventPublisher: EventPublisher,
  ) {}

  async execute(orderId: string, reason: string): Promise<CancelOrderOutput> {
    const order = await this.orderRepository.findById(orderId);
    if (!order) {
      throw new OrderNotFoundError(orderId);
    }

    order.cancel(reason);

    await this.orderRepository.save(order);
    await this.eventPublisher.publishAll([...order.domainEvents]);
    order.clearEvents();

    return {
      orderId: order.id,
      status: order.status,
      reason,
    };
  }
}
